pkg/storagev2/migration/index.go

```diff
@@ -0,0 +1,124 @@
+package migration
+
+import (
+	storage "github.com/ethersphere/bee/pkg/storagev2"
+)
+
+type (
+	// ItemDeleteFn is callback function called in migration step
+	// to check if Item should be removed in this step.
+	ItemDeleteFn func(storage.Item) (delete bool)
+
+	// ItemUpdateFn is callback function called in migration step
+	// to check if Item should be updated in this step.
+	ItemUpdateFn func(storage.Item) (updateditem storage.Item, hasChanged bool)
+)
+
+// WithItemDeleteFn return option with ItemDeleteFn set.
+func WithItemDeleteFn(fn ItemDeleteFn) option {
+	return func(o *options) {
+		o.deleteFn = fn
+	}
+}
+
+// WithItemUpdaterFn return option with ItemUpdateFn set.
+func WithItemUpdaterFn(fn ItemUpdateFn) option {
+	return func(o *options) {
+		o.updateFn = fn
+	}
+}
+
+type option func(*options)
+
+type options struct {
+	deleteFn ItemDeleteFn
+	updateFn ItemUpdateFn
+}
+
+func newOptions() *options {
+	return &options{
+		deleteFn: func(storage.Item) bool { return false },
+		updateFn: func(i storage.Item) (storage.Item, bool) { return i, false },
+	}
+}
+
+func (o *options) applyAll(opts []option) {
+	for _, opt := range opts {
+		opt(o)
+	}
+}
+
+// NewStepOnIndex creates new migration step with update and/or delete operation.
+// Migration will iterate on all elements selected by query and delete or update items
+// based on supplied callback functions.
+func NewStepOnIndex(query storage.Query, opts ...option) StepFn {
+	o := newOptions()
+	o.applyAll(opts)
+
+	return func(s storage.Store) error {
+		itemsForDelete := make([]storage.Item, 0)
+		itemsForUpdate := make([]storage.Item, 0)
+
+		err := s.Iterate(query, func(r storage.Result) (bool, error) {
```

Comment:
We will need to do this in batches. We can potentially have millions of entries which can cause out of memory here.

Check [this](https://github.com/ethersphere/bee/blob/localstorev2.0/pkg/localstorev2/internal/pinning/pinning.go#L248) for example.

Reply:
I am not sure if this example could be used here... In this example iteration continues while there are elements selected by query, so it is possible to stop it at any time and start iteration again.

In case of code from PR, iteration should continue where previous has stopped. Is there a way to achieve such behaviour?

Reply:
So in this case, I would suggest adding support to provide multiple step functions for a particular migration step. This way users can configure different functions for different queries.

Reply:
But then each step is like a version. We dont need to bump the version for every change in a particular item. There might be multiple of these for a particular version bump.

Reply:
> I am not sure if this example could be used here... In this example iteration continues while there are elements selected by query, so it is possible to stop it at any time and start iteration again.
> 
> In case of code from PR, iteration should continue where previous has stopped. Is there a way to achieve such behaviour?

So what we need to do is maintain a skip count. In the IterateFn we can skip entries till this count and then continue. Also, if you delete entries we need not increment the skip count as the next iteration will not show these.

There is no other way than to do this as we can run into out of memory here if the store is few GBs.